PyInstaller/building/utils.py

```diff
@@ -640,3 +640,29 @@ def strip_paths_in_code(co, new_filename=None):
                      co.co_varnames, new_filename, co.co_name,
                      co.co_firstlineno, co.co_lnotab,
                      co.co_freevars, co.co_cellvars)
+
+
+def fake_pyc_timestamp(buf):
+    """
+    Reset the timestamp from a .pyc-file header to a fixed value.
+
+    This enables deterministic builds without having to set pyinstaller
+    source metadata (mtime) since that changes the pyc-file contents.
+
+    _buf_ must at least contain the full pyc-file header.
+    """
+    current_ts = buf[4:8]
+    start, end = 4, 8
+    if current_ts == b'\x00\x00\x00\x01':
```

Comment:
Fine you took care about the upcoming format. But I'm afraid, the detection mechanism is not reliable. Some other tool may have manipulated the timestamp and then this code will destroy the bytecode.

The normal behavior would be to compare the magic which determines the file-format. Since PEP 552 is not yet implemented, we don't know the magic to test for. Anyway, since the new format may only occur with Python >= 3.7, it should be okay to run code-branch `if compat.py_37`. (Since the code has not yet landed, I suggest using `if False and compat.py_37: # TODO py_27`, to make the nightly builds still pass. And we don't care about alpha-versions later on.)

Also this 32-bit word [is defined](https://www.python.org/dev/peps/pep-0552/#specification) to be a bit field. For now PEP 552 only defined the lowest two bits (`hash_based` and `check_source`), so you need to compere the *bit* values. Furter PEPs may define other bits, and the check will fail.

So the code should be something like
```
if False and compat.py_37: # TODO py_37 Python 3.7
   (flags,) = struct.unpack_from(">I", buf, 4)
  If flags & 1:
    # hash-based, clear "check_source" flag, since there is no source
    buf[4:8] = strict.pack(">I", flags ^ 2)
  else:
    # timestamp-base, clear timestamp
    …
else:
  # old format
 ...